tools/export/cmake/__init__.py

```diff
@@ -0,0 +1,242 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import re
+import shutil
+from os import remove
+from os.path import splitext, basename, exists
+from subprocess import Popen, PIPE
+
+from jinja2.exceptions import TemplateNotFound
+
+from tools.export.exporters import Exporter, apply_supported_whitelist
+from tools.targets import TARGET_MAP
+from tools.utils import NotSupportedException
+
+
+class CMake(Exporter):
+    """Generic CMake template that mimics the behavior of the python build
+    system
+    """
+
+    TEMPLATE = 'CMakeLists.txt'
+
+    MBED_CONFIG_HEADER_SUPPORTED = True
+
+    PREPROCESS_ASM = False
+
+    POST_BINARY_WHITELIST = set([
+        "MCU_NRF51Code.binary_hook",
+        "TEENSY3_1Code.binary_hook",
+        "LPCTargetCode.lpc_patch",
+        "LPC4088Code.binary_hook"
+    ])
+
+    @classmethod
+    def is_target_supported(cls, target_name):
+        target = TARGET_MAP[target_name]
+        return apply_supported_whitelist(
+            cls.TOOLCHAIN, cls.POST_BINARY_WHITELIST, target)
+
+    def generate(self):
+        """Generate the CMakefiles.txt
+        """
+        self.resources.win_to_unix()
+
+        # get all source files including headers, adding headers allows IDEs to detect which files
+        # belong to the project, otherwise headers may be greyed out and not work with inspection
+        # (that is true for CLion and definitely for Visual Code)
+        allSourceFiles = set(self.resources.c_sources +
+                             self.resources.cpp_sources +
+                             self.resources.s_sources +
+                             self.resources.headers)
+
+        # create a list of dependencies (mbed add ...)
+        dependencies = [l[:-4] for l in self.resources.lib_refs]
+        # separate the individual dependency source files into a map with the dep name as key and an array if files
+        depSources = {re.sub(r'^[.]/', '', l):
+                          sorted([f for f in allSourceFiles if f.startswith(l)]) for l in dependencies}
+        # delete dependencies that have no source files (may happen if a sub-dependency is ignored by .mbedignore)
+        depSources = {k: v for k, v in depSources.items() if len(v) != 0}
+
+        # remove all source files that ended up being part of one of the dependencies
+        # we flatten the list of source files from all dependencies and
+        # then only add file to srcs if its not in that list
+        # (basically srcs = allSourcefiles - flatten(depSources.values())
+        srcs = [f for f in allSourceFiles if f not in [item for sublist in depSources.values() for item in sublist]]
```

Comment:
I think this may be easier with a set. For example:
```python
srcs = set(allSourceFiles) - sum(set(dep) for dep in depSources.values(), set())
```

I think you mentioned something about python lists not supported subtraction. Python sets do. Maybe this was the logic you were looking for.

Reply:
I found an even simpler and more readable way with your hint.